state_duration.go

```diff
@@ -0,0 +1,138 @@
+package kapacitor
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/influxdata/kapacitor/models"
+	"github.com/influxdata/kapacitor/pipeline"
+	"github.com/influxdata/kapacitor/tick/stateful"
+)
+
+type StateDurationNode struct {
+	node
+	sd         *pipeline.StateDurationNode
+	stateStart time.Time
+
+	expressions map[models.GroupID]stateful.Expression
+	scopePools  map[models.GroupID]stateful.ScopePool
+}
+
+func newStateDurationNode(et *ExecutingTask, sd *pipeline.StateDurationNode, l *log.Logger) (*StateDurationNode, error) {
+	if sd.Lambda == nil {
+		return nil, fmt.Errorf("nil expression passed to StateDurationNode")
+	}
+
+	sdn := &StateDurationNode{
+		node: node{Node: sd, et: et, logger: l},
+		sd:   sd,
+
+		expressions: make(map[models.GroupID]stateful.Expression),
+		scopePools:  make(map[models.GroupID]stateful.ScopePool),
+	}
+	sdn.node.runF = sdn.runStateDuration
+	return sdn, nil
+}
+
+func (sdn *StateDurationNode) expression(g models.GroupID) (stateful.Expression, error) {
+	expr := sdn.expressions[g]
+	if expr == nil {
+		var err error
+		expr, err = stateful.NewExpression(sdn.sd.Lambda.Expression)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to compile expression in stateDuration: %v", err)
+		}
+		sdn.expressions[g] = expr
+	}
+	return expr, nil
+}
+
+func (sdn *StateDurationNode) scopePool(g models.GroupID) stateful.ScopePool {
+	scopePool := sdn.scopePools[g]
+	if scopePool == nil {
+		scopePool = stateful.NewScopePool(stateful.FindReferenceVariables(sdn.sd.Lambda.Expression))
+		sdn.scopePools[g] = scopePool
+	}
+	return scopePool
+}
+
+func (sdn *StateDurationNode) runStateDuration(_ []byte) error {
+	switch sdn.Provides() {
+	case pipeline.StreamEdge:
+		for p, ok := sdn.ins[0].NextPoint(); ok; p, ok = sdn.ins[0].NextPoint() {
+			sdn.timer.Start()
+			expr, err := sdn.expression(p.Group)
+			if err != nil {
+				return err
+			}
+
+			scopePool := sdn.scopePool(p.Group)
+
+			pass, err := EvalPredicate(expr, scopePool, p.Time, p.Fields, p.Tags)
+			if err != nil {
+				sdn.logger.Println("E! error while evaluating expression:", err)
+				continue
+			}
+
+			if pass {
+				if sdn.stateStart.IsZero() {
```

Comment:
Should a stateStartTime be recorded for each group? This way if you group by host the output is the duration that each host has been in the state. 

If you want the duration that any host has been in the state then do not groupBy host.